compiler/rustc_symbol_mangling/src/typeid/typeid_itanium_cxx_abi.rs

```diff
@@ -0,0 +1,534 @@
+        ty::Tuple(_) if ty.is_unit() => {
+            typeid.push('v');
+        }
+
+        // Sequence types
+        ty::Tuple(tys) => {
+            // u5tupleI<element-type1..element-typeN>E as vendor extended type
+            let mut s = String::from("u5tupleI");
+            for ty in tys.iter() {
+                s.push_str(&encode_ty(tcx, ty, dict, options));
+            }
+            s.push('E');
+            compress(dict, (ty, TyQ::None), &mut s);
+            typeid.push_str(&s);
+        }
+
+        ty::Array(ty0, len) => {
+            // A<array-length><element-type>
+            let mut s = String::from("A");
+            s.push_str(&format!("{}", &len.val().try_to_scalar().unwrap().to_i64().unwrap()));
+            s.push_str(&encode_ty(tcx, *ty0, dict, options));
+            compress(dict, (ty, TyQ::None), &mut s);
+            typeid.push_str(&s);
+        }
+
+        ty::Slice(ty0) => {
+            // u5sliceI<element-type>E as vendor extended type
+            let mut s = String::from("u5sliceI");
+            s.push_str(&encode_ty(tcx, *ty0, dict, options));
+            s.push('E');
+            compress(dict, (ty, TyQ::None), &mut s);
+            typeid.push_str(&s);
+        }
+
+        // User-defined types
+        ty::Adt(adt_def, substs) => {
+            let mut s = String::new();
+            let def_id = adt_def.0.did;
+            let name = format!("{}", tcx.item_name(def_id));
+            let crate_name = format!("{}", tcx.crate_name(def_id.krate));
+            // FIXME(rcvalle): Identify C char and integer type uses and encode them with their
+            // respective builtin type encoding as specified by the Itanium C++ ABI if
+            // encoding C types is enabled.
+            if name == "c_void"
+                && (crate_name == "core" || crate_name == "std" || crate_name == "libc")
+            {
+                // C char, integer, and void types
+                s.push('v');
+            } else if adt_def.repr().c() {
+                // For for cross-language CFI support, the encoding must be compatible at the FFI
+                // boundary. For instance:
+                //
+                //     struct type1 {};
+                //     void foo(struct type1* bar) {}
+                //
+                // is encoded as:
+                //
+                //     !1 = !{i64 0, !"_ZTSFvP5type1E"}
+                //
+                // So, encode any repr(C) user-defined type [or extern types (i.e., ty::Foreign)] as
+                // <length><name>, where <name> is <unscoped-name>.
+                s.push_str(&format!("{}", name.len()));
+                s.push_str(&name);
+                compress(dict, (ty, TyQ::None), &mut s);
+            } else if adt_def.is_struct() && adt_def.repr().transparent() {
+                // <element-type1>
+                let mut tys = adt_def.all_fields().map(|field| field.ty(tcx, substs));
+                let ty0 = tys.next().unwrap();
```

Comment:
As mentioned earlier by bjorn3, this should use the non-ZST field (see for example `transparent_newtype_field` function):

```rust
use std::marker::PhantomData;

#[repr(transparent)]
pub struct S<'a> {
    a: (),
    b: PhantomData<&'a mut String>,
    c: usize,
}

#[repr(transparent)]
pub enum E<'a> {
    V((), PhantomData<&'a mut String>, usize),
}

pub fn f<'a>(a: S<'a>, b: E<'a>) {}
```

Reply:
I see. Thanks for pointing it out! I'll fix it.

Reply:
Done.